src/rez/package_test.py

```diff
@@ -158,45 +220,106 @@ def print_header(txt, *nargs):
+            # requires=['foo'] would add this requirement to every test env.
+            # This is a requirement addition, whereas on_variants.requires is a
+            # variant selection mechanism.
+            #
+            if isinstance(on_variants, dict) and "requires" in on_variants:
+                reqlist = RequirementList(
+                    variant.variant_requires + on_variants["requires"])
 
-            # expand refs like {root} in commands
-            if isinstance(command, basestring):
-                command = variant.format(command)
-            else:
-                command = map(variant.format, command)
+                if reqlist.conflict:
+                    continue
+
+                # test if variant requires is a subset of on_variants.requires.
+                # This works because RequirementList merges requirements.
+                #
+                if RequirementList(variant.variant_requires) != reqlist:
+                    continue
 
             # show progress
             if self.verbose:
-                print_header(
+                self._print_header(
                     "\nTest: %s\nPackage: %s\n%s\n",
-                    test_name, variant.uri, '-' * 80)
+                    test_name, variant.uri, '-' * 80
+                )
+
+            # add requirements to force the current variant to be resolved.
+            # TODO this is not perfect, and will need to be updated when
+            # explicit variant selection is added to rez (this is a new
+            # feature). Until then, there's no guarantee that we'll resolve to
+            # the variant we want, so we take that into account here.
+            #
+            requires.extend(map(str, variant.variant_requires))
 
-            # create test env
-            key = tuple(requires)
-            context = self.contexts.get(key)
+            # create test runtime env
+            context = self._get_context(requires)
 
             if context is None:
-                if self.verbose:
-                    print_header("Resolving test environment: %s\n",
-                                 ' '.join(map(quote, requires)))
+                self._set_test_result(
+                    test_name,
+                    variant,
+                    "skipped",
+                    "The current environment does not meet test requirements"
+                )
+
+                print("(Skipped)")
+                continue
 
-                context = ResolvedContext(package_requests=requires,
-                                          package_paths=self.package_paths,
-                                          buf=self.stdout,
-                                          timestamp=self.timestamp,
-                                          **self.context_kwargs)
+            if not context.success:
+                self._set_test_result(
+                    test_name,
+                    variant,
+                    "skipped",
```

Comment:
I wanted to avoid the ambiguity that would arise from "failed" meaning either that a test has failed, or that its env failed to resolve. Skipping isn't perfect either but it felt 'less bad' than marking the test as failed when it in fact never ran.

Reply:
Umm ok... As I said high in the PR, I would really go with failed. From a test perspective, if a test can't even run, it's failed, not skipped. The  main problem I see with marking it as skipped is that if would hide the fact that the test environment is probably broken. It also makes it ambiguous when it comes to choosing an exit code.

If we take a look at `tox` for example, it will raise loudly if a test environment can't be resolved. Same for maven, gradle, etc.